server/fileupdater.mjs

```diff
@@ -0,0 +1,28 @@
+const VERSION = 2;
+
+export default function FileUpdater(state) {
+  const v = state._meta.version;
+  if(v == VERSION)
+    return state;
+
+  for(const id in state) {
+    const w = state[id];
+
+    if(v == 1 && state[id].type == 'button') {
+      let isFirstSelect = true;
+      for(const operation of w.clickRoutine) {
+        if(operation.func == 'SELECT') {
+          if(operation.mode === undefined && !isFirstSelect)
+            operation.mode = 'add';
+          if(operation.mode == 'set')
+            delete operation.mode;
+          isFirstSelect = false;
```

Comment:
I'm not clear on what isFirstSelect is accomplishing here? Looks to me like for each clickRoutine it prevents updating subsequent 'SELECT' functions without mode set to use mode = 'add'. Won't this cause incorrect conversions?

Reply:
The idea is: if it's the first `SELECT` in the button, the mode doesn't make a difference. That's why I usually don't set the `mode` parameter there when I make a new button. I feel like it's weird to make the conversion add `mode: add` to those though.

Reply:
Is it your intention to leave `fileupdater` in the code for the indefinite future? If you hope to use it for any other incompatible changes that might get made moving forward, then the `VERSION` hack you use to check if the game in question has already been converted doesn't seem like it will work the next time. You might want to make it explicit to `SELECT` by renaming it. OTOH if you expect this file to apply only to this issue, perhaps the file itself should be renamed.

Reply:
If I understand this correctly, `fileupdater` will run each time a room is loaded, even though the changes it makes will be persisted if the room is saved to disk. Is that correct? If so, is it possible to put a bit of metadata in the room's JSON that contains this bit of version information? That would reduce load time (though obviously not by much), and might come in handy if there are other future changes that require a similar update change. (And, if this happens, the `VERSION` tag should be changed to something more specific to this particular issue.)

Reply:
It does. Every state has a `state._meta.version`. The FileUpdater checks if it is already equal to `VERSION` (the latest version of the state format that is expected by the code) and if it is, it doesn't do anything.

With the next change I would increment `VERSION` to `3` and check with `v < 3` if the change has to be applied to the state.

Reply:
That makes sense. I didn't see anything named `state` or `version` in a CURL dump of a room's `JSON`, which made me think that it wasn't persisted. Where is it?

Reply:
Yes, that API removes the version information and is not meant for long term storage somewhere. Might be better to add it. In VTT files you can see the `_meta`. For the room itself the `_meta` also includes the list of games and players etc.

Reply:
OK, that makes sense to me now.